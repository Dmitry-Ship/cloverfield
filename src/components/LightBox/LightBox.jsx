import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { lightBox, middle, next, previous, wrapper } from './LightBox.scss';

export default class LightBox extends Component {
  constructor(props) {
    super(props);
    this.state = {
      index: props.index,
    };
    this.showNext = this.showNext.bind(this);
    this.showPrevious = this.showPrevious.bind(this);
  }

  showNext() {
    if (this.state.index + 1 === this.props.images.length) {
      return this.setState({ index: 0 });
    }
    this.setState({ index: this.state.index + 1 });
  }

  showPrevious() {
    this.setState({ index: this.state.index - 1 });
  }

  render() {
    let wrapperRef;
    const { images } = this.props;
    const { index } = this.state;
    const isLast = images.length === index + 1;
    const isFirst = index === 0;

    const isDataURL = (s) => {
      const regex = /^\s*data:([a-z]+\/[a-z]+(;[a-z\-]+\=[a-z\-]+)?)?(;base64)?,[a-z0-9\!\$\&\'\,\(\)\*\+\,\;\=\-\.\_\~\:\@\/\?\%\s]*\s*$/i;
      if (s.match(regex)) { return s; }
      return `${s}`;
    };

    const handleClick = (e) => {
      if (e.target === wrapperRef) {
        this.props.closeLightBox();
      }
    };

    return (
      <div className={lightBox} onClick={handleClick} ref={node => (wrapperRef = node)}>
        {!isFirst &&
          <img
            className={previous}
            src={isDataURL(images[index - 1].url)}
            alt=""
            onClick={this.showPrevious}
          />}
          <img
            className={middle}
            src={isDataURL(images[index].url)}
            alt=""
            onClick={this.showNext}
          />

        {!isLast &&
          <img
            className={next}
            src={isDataURL(images[index + 1].url)}
            alt=""
            onClick={this.showNext}
          />}
      </div>
    );
  }
}

LightBox.propTypes = {
  images: PropTypes.arrayOf(PropTypes.object).isRequired,
  index: PropTypes.number.isRequired,
  closeLightBox: PropTypes.func.isRequired,
};
