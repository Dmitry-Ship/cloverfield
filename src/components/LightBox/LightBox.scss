.lightBox {
  position: fixed;
  // display: flex;
  // flex-direction: column;
  // justify-content: center;
  // align-items: center;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  height: 100vh;
  z-index: 2;
  background: rgba(black, 0.8);
  // overflow-y: auto;
}

.image {
  position: absolute;
  top: 50%;
  
  cursor: pointer;  
  display: block;
  max-width: 70%;
  max-height: 70%;
}

.middle {
  composes: image; 
  transform: translate(-50%, -50%); 
  left: 50%;
  
}

.next {
  composes: image;
  transform: translate(100%, -50%); 
  right: 5%;
}

.previous {
  composes: image;
  left: 5%;
  transform: translate(-100%, -50%); 
  
}