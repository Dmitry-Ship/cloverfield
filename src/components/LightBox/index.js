export { default } from './LightBox.jsx';

