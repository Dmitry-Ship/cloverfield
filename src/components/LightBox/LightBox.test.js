import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, mount } from 'enzyme';
import LightBox from './LightBox';

describe('<LightBox />', () => {
  const closeLightBox = jest.fn();
  const images = [
    { id: '1', url: 'image.jpg' },
    { id: '2', url: 'photo.jpg' },
    { id: '3', url: 'file.jpg' },
  ];

  it('should render correctly', () => {
    const tree = renderer.create(
      <LightBox images={images} index={1} closeLightBox={closeLightBox} />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // it('should call closeLightBox', () => {
  //   const wrapper = mount(
  //     <LightBox images={images} index={1} closeLightBox={closeLightBox} />
  //   );

  //   wrapper.find('img').at(1).simulate('click');

  //   const tree = renderer.create(
  //     <LightBox images={images} index={1} closeLightBox={closeLightBox} />
  //   ).toJSON();
  //   expect(tree).toMatchSnapshot();
  // });

  it('should call closeLightBox', () => {
    const wrapper = mount(
      <LightBox images={images} index={1} closeLightBox={closeLightBox} />
    );

    wrapper.find('div').first().simulate('click');

    expect(closeLightBox).toHaveBeenCalled();
  });
});
