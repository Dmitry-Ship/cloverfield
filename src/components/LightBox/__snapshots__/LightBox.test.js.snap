// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LightBox /> should render correctly 1`] = `
<div
  className="lightBox"
  onClick={[Function]}
>
  <img
    alt=""
    className="previous"
    onClick={[Function]}
    src="image.jpg"
  />
  <img
    alt=""
    className="middle"
    onClick={[Function]}
    src="photo.jpg"
  />
  <img
    alt=""
    className="next"
    onClick={[Function]}
    src="file.jpg"
  />
</div>
`;
